compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -85,39 +89,92 @@ chooseDataTiledMMAAttr(TypeRange eTypes, IREE::GPU::TargetAttr target,
-  const int kNumSubgroups = 4;
-  // Number of register space bits to use for accumulators. Should typically be
-  // between 50% and 80% of total available register space, as the accumulator
-  // tends to be larger than the A and B matrix tiles.
-  const int kMaxAccumulatorRegisterBits = 256 * 32;
+  auto wgp = target.getWgp();
+  if (!wgp.getMaxLoadInstructionBits() || !wgp.getVgprSpaceBits() ||
+      !wgp.getWorkgroupSimds()) {
+    // Missing worgroup parameters: data tiling not supported on this target.
+    return std::nullopt;
+  }
 
   MMAAttr intrinsicMma = MMAAttr::get(ctx, *intrinsic);
   auto [intrinsicA, intrinsicB, intrinsicC] = intrinsicMma.getABCVectorTypes();
   // The unrollK factor serves to allow loads from the A and B matrices to use
   // the target ISA's vector loads. For instance, if the ISA has 128-bit loads
   // and each intrinsic consumes only 32 bits from A and B, then we want to set
   // unrollK=4 to turn 4 separate 32-bit loads into one 128-bit load.
-  const int unrollK =
-      kLoadInstructionBits /
-      std::min(
-          intrinsicA.getElementTypeBitWidth() * intrinsicA.getNumElements(),
-          intrinsicB.getElementTypeBitWidth() * intrinsicB.getNumElements());
+  const int unrollK = std::max(
+      1, static_cast<int>(*wgp.getMaxLoadInstructionBits() /
+                          std::min(intrinsicA.getElementTypeBitWidth() *
+                                       intrinsicA.getNumElements(),
+                                   intrinsicB.getElementTypeBitWidth() *
+                                       intrinsicB.getNumElements())));
+
   // The total amount of unrolling along the M and N dimensions is normally
   // limited only by the number of available registers, since larger M and N
   // yields higher arithmetic intensity. Here, we do not yet distinguish between
   // plain unrolling (more instructions on each thread) and
-  // unrolling-to-subgroups (more threads).
-  const int totalUnrollMN =
-      kMaxAccumulatorRegisterBits /
-      (intrinsicC.getElementTypeBitWidth() * intrinsicC.getNumElements());
-  const int totalUnrollM = static_cast<int>(
-      std::floor(std::sqrt(static_cast<float>(totalUnrollMN))));
-  const int totalUnrollN = totalUnrollMN / totalUnrollM;
+  // unrolling-to-subgroups (more threads), since expanding to more subgroups
+  // correspondingly divides the available register space between this many
+  // subgroups, making it cancel out of the equation here.
+  //
+  // We need to solve for two variables here, unroll_m and unroll_n, constrained
+  // by one second-degree polynomial equation expressing that the A, B and C
+  // tiles must fit in VGPR space. Since we have only 1 constraint for two
+  // variables, we self-impose a second constraint for now: that the unrolling
+  // shape should be square, i.e. unrollM == unrollN. Now we have only one
+  // variable, call it x, to solve for.
```

Comment:
This will break down for narrow M or narrow N cases. Perhaps we can impose additional constraints in the event that M or N are narrow. For example, when M is small enough that the solution for unrollM == unrollN is too large, then we can impose an additional constraint on unrollM to be equal to the largest possible unrolling factor, and solve for unrollN:
```
x * M_max * sizeInBits(intrinsicC) + M_max * unrollK * sizeInBits(intrinsicA) + x * unrollK * sizeInBits(intrinsicB) == wgp.getVgprSpaceBits()
```